tools/codegen/gen_vmap_plumbing2.py

```diff
@@ -0,0 +1,213 @@
+    Argument, Return, SchemaKind, Tag
+)
+from tools.codegen.context import method_with_native_function
+from tools.codegen.utils import mapMaybe
+from dataclasses import dataclass
+from typing import List, Dict, Optional, Tuple, Set, Any, Union, Sequence, TypeVar
+import textwrap
+
+
+def is_tensor(typ: Type) -> bool:
+    return isinstance(typ, BaseType) and typ.name == BaseTy.Tensor
+
+
+def is_optional_tensor(typ: Type) -> bool:
+    return isinstance(typ, OptionalType) and is_tensor(typ.elem)
+
+
+def is_vector_tensor(typ: Type) -> bool:
+    return isinstance(typ, ListType) and is_tensor(typ.elem)
+
+
+def unwrap_tensor(name: str) -> List[str]:
+    result = f"""\
+    Tensor {name}_value;
+    optional<int64_t> {name}_bdim;
+    std::tie({name}_value, {name}_bdim) = unwrapTensorAtLevel({name}, cur_level);"""
+    return textwrap.dedent(result).split('\n')
+
+
+def unwrap_optional_tensor(name: str) -> List[str]:
+    result = f"""\
+    optional<Tensor> {name}_value;
+    optional<int64_t> {name}_bdim;
+    if ({name}) {{
+        std::tie({name}_value, {name}_bdim) = unwrapTensorAtLevel({name}.value(), cur_level);
+    }}"""
+    return textwrap.dedent(result).split('\n')
+
+
+def gen_unwraps(flat_arguments: List[Argument]) -> Tuple[List[str], List[str]]:
+    arg_names = [a.name for a in flat_arguments]
+    arg_types = [a.type for a in flat_arguments]
+
+    tensors = [name for typ, name in zip(arg_types, arg_names) if is_tensor(typ)]
+    optional_tensors = [name for typ, name in zip(arg_types, arg_names) if is_optional_tensor(typ)]
+
+    unwraps = []
+    for tensor in tensors:
+        unwraps += unwrap_tensor(tensor)
+
+    for opt_tensor in optional_tensors:
+        unwraps += unwrap_optional_tensor(opt_tensor)
+    unwraps = '\n'.join(unwraps)
+
+    unwrapped_arg_list = []
+    for arg in arg_names:
+        if arg in tensors or arg in optional_tensors:
+            unwrapped_arg_list += [f'{arg}_value', f'{arg}_bdim']
+        else:
+            unwrapped_arg_list.append(arg)
+    return unwraps, unwrapped_arg_list
+
+
+def get_aten_op_call(schema) -> str:
+    if schema.name.overload_name:
+        return f'ATEN_FN2({schema.name.name}, {schema.name.overload_name})'
+    return f'ATEN_FN({schema.name.name})'
+
+
+def gen_case_where_all_bdims_are_none(flat_args: List[Argument], schema) -> str:
+    conditions = []
+    for arg in flat_args:
+        if not arg.type.is_tensor_like():
+            continue
+        conditions.append(f'!isBatchedAtLevel({arg.name}, cur_level)')
+    aten_op = get_aten_op_call(schema)
+    arg_names = [a.name for a in flat_args]
+
+    return f"""\
+if ({' && '.join(conditions)}) {{
+  return {aten_op}({', '.join(arg_names)});
+}}"""
+
+
+def gen_returns(returns: List[Return]) -> str:
+    idx = 0
+    wrapped_returns = []
+    for ret in returns:
+        if is_tensor(ret.type):
+            wrapped_returns.append(f'makeBatched(std::get<{idx}>(results), std::get<{idx + 1}>(results), cur_level)')
```

Comment:
It would be nice if we could avoid hardcoding intermediate variable names in multiple places (`results` here, and I guess `cur_level`)